tests/dragonfly/__init__.py

```diff
@@ -1,11 +1,78 @@
 import pytest
 
+import time
+import subprocess
+
+
+class DflyInstance:
+    """
+    Represents a runnable and stoppable Dragonfly instance
+    with fixed arguments.
+    """
+
+    def __init__(self, path, args, cwd):
+        self.path = path
+        self.args = args
+        self.cwd = cwd
+
+    def start(self):
+        arglist = DflyInstance.format_args(self.args)
+
+        print(f"Starting instance on {self.port()} with arguments {arglist}")
+        self.proc = subprocess.Popen([self.path, *arglist], cwd=self.cwd)
+
+        # Give Dragonfly time to start and detect possible failure causes
+        time.sleep(0.3)
+
+        return_code = self.proc.poll()
+        if return_code is not None:
+            raise Exception(
+                f"Failed to start instance, return code {return_code}")
+
+    def stop(self):
+        print(f"Stopping instance on {self.port()}")
+        try:
+            self.proc.terminate()
+            outs, errs = self.proc.communicate(timeout=15)
+        except subprocess.TimeoutExpired:
+            print("Unable to terminate DragonflyDB gracefully, it was killed")
+            outs, errs = self.proc.communicate()
+            print(outs, errs)
+
+    def port(self) -> int:
+        return int(self.args.get("port", "6379"))
+
+    @staticmethod
+    def format_args(args):
+        out = []
+        for (k, v) in args.items():
+            out.extend((s for s in ("--"+k, v) if s != ""))
+        return out
+
+
+class DflyInstanceFactory:
+    """
+    A factory for creating dragonfly instances with pre-supplied arguments.
+    """
+
```

Comment:
This class might look a bit dull, but I'll extend it for replication testing with context utilities (to not forget stopping instances)

Reply:
To setup and cleanup you have fixture in pytest to making sure to do cleanup - 
for example 
```
@pytest.fixture()
def resource():
    print("setup")
    yield "resource"
    print("teardown")
```
Note that the action after the yield here would do the cleanup after the test (in this case it just print). So I don't think you really in a class in this case.

Reply:
Sure, this is what happens in `df_server` for example. Its not needed here, _but_

I need a way of spawning df instances in my replication PR. Instead of doing all the refactoring there, I instead opened this PR to introduce the changes separately.